### PR TITLE
[14.0][MIG] sale_backorder

### PR DESCRIPTION
**Pull request by Freni-OSI**

### CODE REVIEW

**Comment by patrickrwilson** [comment]
@Freni-OSI can you increase test coverage please?

**Review by patrickrwilson — APPROVED** [review]
Code review: I have no suggestions, looks good. LGTM

**Review by cbeddies — APPROVED** [review]

File: sale_backorder/models/sale.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        string="UIGD Value", compute="_compute_uigd_value", store=True
+    )
+    bo_value = fields.Monetary(
+        string="Backorder Value", compute="_compute_bo_value", store=True
+    )
+
+    @api.depends("order_line.uigd_qty", "order_line.price_unit")
+    def _compute_uigd_value(self):
+        for order in self:
+            total = 0
```
Comment (rousseldenis) [inline sale_backorder/models/sale.py]: `order.uigd_value = sum(order.order_line.mapped("uigd_value")`

File: sale_backorder/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        string="Last Bill Date", compute="_compute_last_bill_date", store=True
+    )
+    uigd_value = fields.Monetary(
+        string="UIGD Value", compute="_compute_uigd_value", store=True
+    )
+    bo_value = fields.Monetary(
+        string="Backorder Value", compute="_compute_bo_value", store=True
+    )
+
+    @api.depends("order_line.uigd_qty", "order_line.price_unit")
```
Comment (rousseldenis) [inline sale_backorder/models/sale.py]:
```suggestion
    @api.depends("order_line.uigd_value")
```
is more correct as you could have cache problems.

File: sale_backorder/models/sale.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    def _compute_uigd_value(self):
+        for order in self:
+            total = 0
+            for line in order.order_line:
+                total += line.uigd_value
+            order.uigd_value = total
+
+    @api.depends("order_line.bo_qty", "order_line.price_unit")
```
Comment (rousseldenis) [inline sale_backorder/models/sale.py]:
```suggestion
    @api.depends("order_line.bo_value")
```

Comment (rousseldenis) [reply sale_backorder/models/sale.py]: same for here under functions

File: sale_backorder/models/sale.py (hunk 1 above)
Comment (rousseldenis) [reply sale_backorder/models/sale.py]: Same for here under functions

File: sale_backorder/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    def _compute_last_bill_date(self):
+        for order in self:
+            max_date = False
+            for line in order.order_line:
+                if line.last_bill_date:
+                    max_date = line.last_bill_date
+            order.last_bill_date = max_date
+
+
+class SaleOrderLine(models.Model):
```
Comment (rousseldenis) [inline sale_backorder/models/sale.py]: You should separate models in distinct files

File: sale_backorder/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    _inherit = "sale.order.line"
+
+    last_date_delivered = fields.Datetime(
+        string="Last Date Delivered", compute="_compute_last_date_delivered", store=True
+    )
+    last_bill_date = fields.Datetime(
+        string="Last Bill Date", compute="_compute_last_bill_date", store=True
+    )
+    uigd_qty = fields.Float(string="UIGD Qty", compute="_compute_uigd_qty", store=True)
```
Comment (rousseldenis) [inline sale_backorder/models/sale.py]: IMHO explicit field names is better

File: sale_backorder/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    @api.depends("bo_qty", "price_unit")
+    def _compute_bo_value(self):
+        for line in self:
+            line.bo_value = line.bo_qty * line.price_unit
+
+    @api.depends("qty_delivered", "qty_invoiced")
+    def _compute_last_date_delivered(self):
+        for line in self:
+            max_date = False
+            for move in line.move_ids:
```
Comment (rousseldenis) [inline sale_backorder/models/sale.py]:
```suggestion
            moves = line.move_ids.filtered(lambda move: move.state != "done" or move.location_dest_id.usage != "customer" or not move.to_refund)
            line.last_date_delivered = first(moves).date
```

Comment (rousseldenis) [reply sale_backorder/models/sale.py]: Same for here under

File: sale_backorder/tests/test_sale_backorder.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
+# Copyright (C) 2021 - TODAY, Open Source Integrators
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from datetime import datetime
+
+from dateutil.relativedelta import relativedelta as rdelta
+
+from odoo.tests.common import TransactionCase
```
Comment (rousseldenis) [inline sale_backorder/tests/test_sale_backorder.py:8]: Better to use SavepointCase

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review.

File: sale_backorder/models/sale.py (hunk 5 above)
Comment (Vandan-OSI) [reply sale_backorder/models/sale.py]: @rousseldenis i.e field name is Like un_Invoiced_goods_delivered ??

Comment (rousseldenis) [reply sale_backorder/models/sale.py]: Put yourself at the place of user/developer that came after.

uninvoiced_goods_delivered_qty. I suppose string is to help display in tree view. But you should add a help on that field too.

**Review by max3903 — APPROVED** [review]

**Comment by max3903** [comment]
@Urvisha-OSI Can you take care of the comments please?

**Comment by Hardik-OSI** [comment]
Hello @max3903 

sale_backorder module all code-coverage 100 % there are some other module which needs to add code coverage.

also for pre-commit i add one issue in this repo there is wrong flake8 url added in pre-commit-config.yml file.

Thanks.

**Comment by francescapenso** [comment]
Should it works also with service-product?
Such as i sell 2 product and invoice just 1? Should i see in UIGD value the difference from amount sell and amount invoiced?

**Pull request merged**